sklearn/ensemble/_hist_gradient_boosting/_predictor.pyx

```diff
@@ -21,85 +21,168 @@ from .common cimport node_struct
 np.import_array()
 
 
-def _predict_from_numeric_data(
-        node_struct [:] nodes,
-        const X_DTYPE_C [:, :] numeric_data,
-        Y_DTYPE_C [:] out):
-
-    cdef:
-        int i
+@cython.final
+cdef class TreePredictor:
+    """Tree class used for predictions.
 
-    for i in prange(numeric_data.shape[0], schedule='static', nogil=True):
-        out[i] = _predict_one_from_numeric_data(nodes, numeric_data, i)
+    Parameters
+    ----------
+    n_nodes : int
+        Number of nodes in tree.
+    """
+    def __init__(self, int n_nodes):
+        self.nodes.resize(n_nodes)
+
+    cdef node_struct* get(self, int node_idx) nogil:
+        """Get reference to node."""
+        return &self.nodes[node_idx]
+
+    cdef int get_n_nodes(self) nogil:
+        """Get number of nodes in tree."""
+        return self.nodes.size()
+
+    def get_n_leaf_nodes(self):
+        """Return number of leaves."""
+        cdef node_struct node
+        return sum(node.is_leaf for node in self.nodes)
+
+    def get_max_depth(self):
+        """Return maximum depth among all leaves."""
+        cdef node_struct node
+        return max(node.depth for node in self.nodes)
+
+    def predict(self, const X_DTYPE_C [:, :] X):
+        """Predict raw values for non-binned data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input samples.
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples,)
+            The raw predicted values.
+        """
+        cdef:
+            int i
+            Y_DTYPE_C [:] out = np.empty(X.shape[0], dtype=Y_DTYPE)
+
+        for i in prange(X.shape[0], schedule='static', nogil=True):
+            out[i] = _predict_one_from_numeric_data(self, X, i)
+        return np.asarray(out)
+
+    def predict_binned(self, const X_BINNED_DTYPE_C [:, :] X,
+                       const unsigned char missing_values_bin_idx):
+        """Predict raw values for binned data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input samples.
+        missing_values_bin_idx : uint8
+            Index of the bin that is used for missing values. This is the
+            index of the last bin and is always equal to max_bins (as passed
+            to the GBDT classes), or equivalently to n_bins - 1.
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples,)
+            The raw predicted values.
+        """
+        cdef:
+            int i
+            Y_DTYPE_C [:] out = np.empty(X.shape[0], dtype=Y_DTYPE)
+
+        for i in prange(X.shape[0], schedule='static', nogil=True):
+            out[i] = _predict_one_from_binned_data(self, X, i,
```

Comment:
If we're passing `self`, should this just be a method? Same for `_predict_one_from_numeric_data`